### PR TITLE
[Spec] @-based annotation syntax

### PR DESCRIPTION
**Pull request by abreslav**

### CODE REVIEW

**Comment by EddieRingle** [comment]
:+1: I always thought @-style annotations in Java made sense considering "@" is basically just a stylized "A"

Would we still be able to use @-based annotations without any special syntax (i.e., `data` instead of `@data`) on classes?

Side-note: @abreslav Out of curiousity, are you open to hearing language suggestions in the form of Pull Requests here from people outside of JetBrains?

**Comment by abreslav** [comment]
> Would we still be able to use @-based annotations without any special syntax (i.e., `data` instead of `@data`) on classes?
Yes.

> are you open to hearing language suggestions in the form of Pull Requests here from people outside of JetBrains?

For people not on the Kotlin team, it's better to discuss suggestions with us in the [forum](http://devnet.jetbrains.net/community/kotlin) first

File: spec-docs/at-based-annotation-syntax.md [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+```
+
+Downside: to put the same annotation on two targets, we'd need to duplicate it.
+
+For this, we need to allow annotation attributes of type `Array<Annotation>`:
+
+``` kotlin
+annotation class field(vararg val annotations: Annotation)
+```
+**NOTE**: This is only relatively easilty achievable for source-retained annotations, for class- or runtime-retained it's a lot more involved and relies on an undocumented features of JVM.
```
Comment (goodwinnk) [inline spec-docs/at-based-annotation-syntax.md]: eastlty -> easily

File: spec-docs/at-based-annotation-syntax.md [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+Possible targets are
+* `field`
+* `get`
+* `set`
+* `property`
+* `parameter` - for constructor parameters that are also properties
+
+Reasonable defaults would probably be:
+* `field` if there's a backing field
+* `get` otherwise
```
Comment (goodwinnk) [inline spec-docs/at-based-annotation-syntax.md]: Maybe it's worth to mention default target for constructor parameters? Will it be `parameter` for simple parameters and for fields?

File: spec-docs/at-based-annotation-syntax.md [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+Otherwise, determined by the settings of the annotation itself (applicable to fields only -> goes to a field)
```
Comment (goodwinnk) [inline spec-docs/at-based-annotation-syntax.md]: Shall we really allow omitting the non-default target in code? It's a clear message with quick-fix for a code writter and possible source of misunderstanding for a reader that can't be resolved without navigating to annotation declaration. And annotation target settings modification can lead to unexpected tools failures without any compiler messages on rebuild.

**Comment by abreslav** [comment]
Moved to https://gist.github.com/abreslav/6ae8195ebae37ab5219351eb576e0a00

**Pull request closed without merging**